include/nbl/builtin/hlsl/emulated/vector_t.hlsl

```diff
@@ -0,0 +1,523 @@
+struct is_valid_emulated_vector
+{
+    NBL_CONSTEXPR_STATIC bool value = is_same_v<VecType, emulated_vector_t2<ComponentType> > ||
+        is_same_v<VecType, emulated_vector_t3<ComponentType> > ||
+        is_same_v<VecType, emulated_vector_t4<ComponentType> >;
+};
+
+template<typename U, typename T, typename I = uint32_t>
+struct array_get
+{
+    T operator()(NBL_REF_ARG(U) vec, const I ix)
+    {
+        return vec[ix];
+    }
+};
+
+template<typename ComponentType>
+struct array_get<emulated_vector_t2<ComponentType>, ComponentType, uint32_t>
+{
+    ComponentType operator()(NBL_REF_ARG(emulated_vector_t2<ComponentType>) vec, const uint32_t ix)
+    {
+        return vec.getComponent(ix);
+    }
+};
+
+template<typename ComponentType>
+struct array_get<emulated_vector_t3<ComponentType>, ComponentType, uint32_t>
+{
+    ComponentType operator()(NBL_REF_ARG(emulated_vector_t3<ComponentType>) vec, const uint32_t ix)
+    {
+        return vec.getComponent(ix);
+    }
+};
+
+template<typename ComponentType>
+struct array_get<emulated_vector_t4<ComponentType>, ComponentType, uint32_t>
+{
+    ComponentType operator()(NBL_REF_ARG(emulated_vector_t4<ComponentType>) vec, const uint32_t ix)
+    {
+        return vec.getComponent(ix);
+    }
+};
+
+#undef DEFINE_EMULATED_VECTOR_ARRAY_GET_SPECIALIZATION
+
+template<typename U, typename T, typename I = uint32_t>
+struct array_set
+{
+    void operator()(NBL_REF_ARG(U) arr, I index, T val)
+    {
+        arr[index] = val;
+    }
+};
+
+template<typename ComponentType>
+struct array_set<emulated_vector_t2<ComponentType>, ComponentType, uint32_t>
+{
+    void operator()(NBL_REF_ARG(emulated_vector_t2<ComponentType>) vec, uint32_t index, ComponentType value)
+    {
+        vec.setComponent(index, value);
+    }
+};
+
+template<typename ComponentType>
+struct array_set<emulated_vector_t3<ComponentType>, ComponentType, uint32_t>
+{
+    void operator()(NBL_REF_ARG(emulated_vector_t3<ComponentType>) vec, uint32_t index, ComponentType value)
+    {
+        vec.setComponent(index, value);
+    }
+};
+
+template<typename ComponentType>
+struct array_set<emulated_vector_t4<ComponentType>, ComponentType, uint32_t>
+{
+    void operator()(NBL_REF_ARG(emulated_vector_t4<ComponentType>) vec, uint32_t index, ComponentType value)
+    {
+        vec.setComponent(index, value);
+    }
+};
+
+namespace impl
+{
+template<typename To, typename From>
+struct static_cast_helper<emulated_vector_t2<To>, vector<From, 2>, void>
+{
+    static inline emulated_vector_t2<To> cast(vector<From, 2> vec)
+    {
+        emulated_vector_t2<To> output;
+        output.x = _static_cast<To, From>(vec.x);
+        output.y = _static_cast<To, From>(vec.y);
+
+        return output;
+    }
+};
+
+template<typename To, typename From>
+struct static_cast_helper<emulated_vector_t3<To>, vector<From, 3>, void>
+{
+    static inline emulated_vector_t3<To> cast(vector<From, 3> vec)
+    {
+        emulated_vector_t2<To> output;
```

Comment:
`_t2` is this a bug?

Reply:
@Przemog1 is there a reason why a nicer partial spec (that specs against dimensions) didn't work?

Reply:
> @Przemog1 is there a reason why a nicer partial spec (that specs against dimensions) didn't work?

I don't remember ;_;

Reply:
> @Przemog1 is there a reason why a nicer partial spec (that specs against dimensions) didn't work?

![image](https://github.com/user-attachments/assets/a6179b72-6ed6-49b8-9d22-262d0e87319c)
![image](https://github.com/user-attachments/assets/3509b475-1600-474c-8fc1-3a520e801837)
@devshgraphicsprogramming thats why

Reply:
ok, I see